packages/modules/notification/src/services/notification-module-service.ts

```diff
@@ -108,57 +109,114 @@ export default class NotificationModuleService
     const idempotencyKeys = data
       .map((entry) => entry.idempotency_key)
       .filter(Boolean)
-    const alreadySentNotifications = await this.notificationService_.list(
-      {
-        idempotency_key: idempotencyKeys,
-      },
-      { take: null },
-      sharedContext
-    )
 
-    const existsMap = new Map(
-      alreadySentNotifications.map((n) => [n.idempotency_key as string, true])
-    )
+    let { notificationsToProcess, createdNotifications } =
+      await this.baseRepository_.transaction(async (txManager) => {
+        const context = {
+          ...sharedContext,
+          transactionManager: txManager,
+        }
 
-    const notificationsToProcess = data.filter(
-      (entry) => !entry.idempotency_key || !existsMap.has(entry.idempotency_key)
-    )
+        const alreadySentNotifications = await this.notificationService_.list(
+          {
+            idempotency_key: idempotencyKeys,
+          },
+          { take: null },
+          context
+        )
 
-    const notificationsToCreate = await promiseAll(
-      notificationsToProcess.map(async (entry) => {
-        const provider =
-          await this.notificationProviderService_.getProviderForChannel(
-            entry.channel
-          )
+        const existsMap = new Map(
+          alreadySentNotifications.map((n) => [n.idempotency_key as string, n])
+        )
 
-        if (!provider) {
-          throw new MedusaError(
-            MedusaError.Types.NOT_FOUND,
-            `Could not find a notification provider for channel: ${entry.channel}`
-          )
-        }
+        const notificationsToProcess = data.filter(
+          (entry) =>
+            !entry.idempotency_key ||
+            !existsMap.has(entry.idempotency_key) ||
+            (existsMap.has(entry.idempotency_key) &&
+              [NotificationStatus.PENDING, NotificationStatus.FAILURE].includes(
```

Comment:
We probably shouldn't process the notification if it is in pending, since another process might already be processing it, so let's just compare against failures